azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/AzureAccount.java

```diff
@@ -45,128 +47,144 @@ public Account account() throws AzureToolkitAuthenticationException {
+                accounts.addAll(accountByType.values().stream().map(Supplier::get).collect(Collectors.toList()));
+            } else {
+                if (!accountByType.containsKey(type)) {
+                    return Mono.error(new LoginFailureException(String.format("Unsupported auth type '%s', supported values are: %s.\"",
+                            type, accountByType.keySet().stream().map(Object::toString).map(StringUtils::lowerCase).collect(Collectors.joining(", ")))));
+                }
+                accounts.add(accountByType.get(type).get());
+            }
+        }
 
-            String expectedEnv = AzureEnvironmentUtils.getCloudNameForAzureCli(auth.getEnvironment());
-            String realEnv = AzureEnvironmentUtils.getCloudNameForAzureCli(this.account.getEnvironment());
+        return authenticateAccounts(accounts).doOnSuccess(result -> {
+            checkEnv(result, auth.getEnvironment());
+            this.setAccount(result);
+        });
+    }
+
+    public Mono<Account> loginAsync(AuthType type) {
+        Objects.requireNonNull(type, "Please specify auth type in auth configuration.");
+        Map<AuthType, Supplier<Account>> accountByType = buildAccountMap();
+        if (!accountByType.containsKey(type)) {
+            return Mono.error(new LoginFailureException(String.format("Unsupported auth type '%s', supported values are: %s.\"",
+                    type, accountByType.keySet().stream().map(Object::toString).map(StringUtils::lowerCase).collect(Collectors.joining(", ")))));
+        }
+        return loginAsync(accountByType.get(type).get());
+    }
+
+    public Mono<Account> loginAsync(Account targetAccount) {
+        Objects.requireNonNull(targetAccount, "Please specify account to login.");
+
+        return targetAccount.authenticate().doOnSuccess(this::setAccount);
+    }
+
+    private AzureAccount blockMonoAndReturnThis(Mono<Account> mono) {
+        try {
+            mono.block();
+            return this;
+        } catch (Throwable ex) {
+            throw new AzureToolkitAuthenticationException("Cannot login due to error: " + ex.getMessage());
+        }
+    }
+
+    private static void checkEnv(Account ac, AzureEnvironment env) {
+
+        if (env != null && ac.getEnvironment() != null
+                && ac.getEnvironment() != env
+                && ac.isAuthenticated()) {
+
+            String expectedEnv = AzureEnvironmentUtils.getCloudNameForAzureCli(env);
+            String realEnv = AzureEnvironmentUtils.getCloudNameForAzureCli(ac.getEnvironment());
 
             // conflicting configuration of azure environment
-            switch (this.account.getMethod()) {
+            switch (ac.getMethod()) {
                 case AZURE_CLI:
-                    throw new LoginFailureException(String.format("The azure cloud from azure cli '%s' doesn't match with your auth configuration, " +
+                    throw new AzureToolkitAuthenticationException(
```

Comment:
As env strategy differs from different account, shall we handle the env check in Account?

Reply:
It will make account more complex with two sources of env, keep it here is better